### PR TITLE
Weight generation for integrated bilinear generalized barycentric schemes for meshes with holes

### PR DESCRIPTION
**Pull request by dmarsico1**
Added weight generation for integrated bilinear and generalized barycentric schemes for meshes with holes (the methods “intbilin” and “intbilingb”)

In order to have a consistent map when the source mesh has holes, the option “normalize” must be used (this is due to the target faces near coastlines being only partially covered by source faces).   So consistent map generation for the method "intbilin" and "intbilingb" when the source mesh has holes would be something like this:

./GenerateOfflineMap --in_mesh mesh_in.g --out_mesh mesh_out.g \
                     --ov_mesh mesh_overlap.g --method “intbilin;normalize” \
                     --out_map map.nc

I also added an integrated nearest neighbor routine, and edited the (non-integrated) bilinear and generalized barycentric coordinate schemes to remove an unnecessary conditional.  

I added a few lines to the function used to calculate the generalized barycentric coordinates that is necessary to account for the instability caused when the input face has too many edges.

### CODE REVIEW

**Comment by vijaysm** [comment]
Comment: Isn't it easy to figure out if the mesh has holes? Why does the user remember to pass the `normalize` option? More often than not, this will get forgotten and will just end up as an issue. On the other hand, does adding the `normalize` option for global meshes affect the generated map? If it does not, then perhaps just make `normalize` as default ON.

**Comment by vijaysm** [comment]
Also, in general it is better to rebase your branch against origin/master. It makes history cleaner and linear.

**Comment by vijaysm** [comment]
Another question: do the `int` in `intbilin` and `intbilingb` refer to integrated? Does that mean I should expect to see normalizations to preserve the global integral on the source mesh? The reason I ask is because for conservative FV on meshes with holes, I get for example 
```
..   Consist. err min/max: -1.000000000000000e+00 / 2.220446049250313e-16
..    Col wt.sums min/max: 0.000000000000000e+00 / 1.000000000000001e+00
..   Conserv. err min/max: -1.000000000000000e+00 / 6.661338147750939e-16
..    Global conserv. err: 0.000000000000000e+00
```
and with `intbilingb;normalize`, I get
```
..   Consist. err min/max: -1.000000000000000e+00 / 2.220446049250313e-16
..    Col wt.sums min/max: 0.000000000000000e+00 / 1.026034470160455e+00
..   Conserv. err min/max: -1.000000000000000e+00 / 2.603447016045490e-02
..    Global conserv. err: 2.925853279802357e-01
```
Is this expected? Please clarify if I have misunderstood what `int*` is supposed to be doing. @paullric @dmarsico1

File: src/FiniteVolumeTools.cpp [hunk 1]
```diff
@@ -1673,6 +1673,44 @@ void GeneralizedBarycentricCoordinates(
 			
 		}
 		
+		//Use nearest neighbor weighting if there are too many edges.  This is necessary to avoid numerical instability.  
+		
+		if( iEdges > 30 ){
```
Comment (vijaysm) [inline src/FiniteVolumeTools.cpp:1678]: Should I also test with a coarse CS and a fine RLL mesh to trigger this case? What is the resolution combo that you tested for this edge case?

File: src/LinearRemapFV.cpp [hunk 2]
```diff
@@ -1315,272 +1314,460 @@ void LinearRemapIntegratedGeneralizedBarycentric(
...
+			int nOverlapFaces = ixOverlapEnd - ixOverlapBegin;
+	
+			if( nOverlapFaces == 0 ) continue;
+			
+			std::vector<std::vector<int>> vecContributingFaceI;
+			
+			std::vector<std::vector<double>> vecContributingWeights;		
+	
+			// Loop through all overlap faces associated with this source face
+			for (int j = 0; j < nOverlapFaces; j++) {
```
Comment (vijaysm) [inline src/LinearRemapFV.cpp:1560]: Perhaps a point for later: if bilinear family and barycenter integrated schemes use the same workflow for mesh with and without holes, see opportunities to refactor and unify under a similar structure. If there is a bug, there are a lot of places the same code needs to be modified.

File: src/LinearRemapFV.cpp [hunk 3]
```diff
@@ -2662,164 +2825,584 @@ void LinearRemapIntegratedBilinear(
...
+						GetFaceThatContainsPoint(meshInputDual,iNearestFace,iFaceFinal,nodeQ.x,nodeQ.y,nodeQ.z);
+						
+						int iEdges = meshInputDual.faces[iFaceFinal].edges.size();
+						
+						//NodeVector nodesCurrentFace(iEdges);
+						NodeVector nodesCurrentFace;
+						
+						for (int i = 0; i < iEdges; i++){
+							
+							nodesCurrentFace.push_back(meshInputDual.nodes[meshInputDual.faces[iFaceFinal][i]]);
```
Comment (vijaysm) [inline src/LinearRemapFV.cpp:3032]: The diffs are absolutely terrible in the interface. The local git diff isn't any better unfortunately, which makes it very hard to track new code changes in the `LinearRemapIntegratedBilinear` routine. As far as my review goes, I don't see anything significantly out of the ordinary in the new additions, but if there is a specific chunk of code I should review, let me know.

**Review by vijaysm — CHANGES_REQUESTED** [review]

File: src/FiniteVolumeTools.cpp (hunk 1 above)
Comment (dmarsico1) [reply src/FiniteVolumeTools.cpp:1678]: This only occurs for one specific land source mesh with triangular elements where there is a mesh node that is shared by 720 source faces (it occurs near the south pole, all the other nodes are fine). The dual mesh face that surrounds this node then contains 720 edges, which leads to extreme ill-condition of the algorithm that is used to generated the barycentric coordinates.  There are no RLL or CS meshes of any resolution that will lead to this problem.  Note that the bilinear weighting does not face this problem because faces that contain more than 4 edges are triangulated.

I can send you this mesh if you want to test it.  Let me know.

File: src/LinearRemapFV.cpp (hunk 2 above)
Comment (dmarsico1) [reply src/LinearRemapFV.cpp:1560]: Got it.  I will work on unifying the code here and submit in another pull request.

File: src/LinearRemapFV.cpp (hunk 3 above)
Comment (dmarsico1) [reply src/LinearRemapFV.cpp:3036]: Most of the changes are in the functions LinearRemapIntegratedBilinear and LinearRemapIntegratedGeneralizedBarycentric.  Originally, these functions only worked for global source meshes, so they had to be altered significantly to account for source meshes with holes.  Many of the other changes involved simply replacing blocks of code with calls to functions.  For instance, the bilinear and barycentric weights are now obtained by calling the functions “BilinearWeights” and “GeneralizedBarycentricCoordinates,” respectively.  

For the most part the LinearRemapIntegratedBilinear and LinearRemapIntegratedGeneralizedBarycentric functions are similar to their non-integrated counterparts in LinearRemapBilinear and LinearRemapGeneralizedBarycentric, apart from the fact that the integrated versions loop over the overlap mesh, instead of the target mesh.

**Comment by dmarsico1** [comment]
> Comment: Isn't it easy to figure out if the mesh has holes? Why does the user remember to pass the `normalize` option? More often than not, this will get forgotten and will just end up as an issue. On the other hand, does adding the `normalize` option for global meshes affect the generated map? If it does not, then perhaps just make `normalize` as default ON.

The normalization of rows has now been added directly to the code so that the integrated schemes enforce consistency by default when the source mesh has holes.  The “normalize” parameter is no longer needed.

File: src/FiniteVolumeTools.cpp (hunk 1 above)
Comment (vijaysm) [reply src/FiniteVolumeTools.cpp:1678]: Thanks for sending me the relevant meshes.

File: src/LinearRemapFV.cpp (hunk 3 above)
Comment (vijaysm) [reply src/LinearRemapFV.cpp:3036]: Thanks for the clarifications. I reviewed the code again and the changes are similar to the non-integrated variants as you mentioned.

**Comment by vijaysm** [comment]
> Yes, the “int" in intbilin and intbilingb refers to integrated. Despite the name, however, the integrated schemes do not conserve global integrals. They’re referred to as integrated because the variables are treated as cell averages (even though they’re not conserved). Approximating the variables in this manner allows every point on the source mesh to be sampled and eliminates the significant error that arises when the source mesh is finer than the target mesh. (See our GMD paper “Strategies for Conservative and Non-Conservative monotone remapping on the sphere” for more details).

I see. I did read the paper but perhaps the details eluded me now. Thanks for the clarification. The output makes sense now.

**Review by vijaysm — APPROVED** [review]
The requested changes and comments have been adequately addressed. A follow-up PR to consolidate and refactor bilin* variant changes can help simplify code readability and maintainability. But for now, the changes are good to go.
